### ISSUE TITLE
Feature proposal: control what is shown in the OSD lists of cores, disk images, etc.

### ISSUE THREAD

**Issue by Frank-Zeyda** [MiSTer-devel/Main_MiSTer#443]
Dear MiSTer Developers & Users,

Hope you are all well and healthy in these strange times.

I would like to propose a new feature that gives users a little more control over what is shown in the various OSD file lists, i.e., for arcade cores, disk images, etc., namely when entering a given directory.

For this, I suggest a .showlist and .hidelist configuration file, to be written to the respective directory. These are optional files and, if not present, do not alter the current business logic of the OSD in terms of showing **all** applicable files in the active directory.

If a .showlist is provided, only files that match an entry (line) in the .showlist are displayed. If a .hidelist is provided, files that match one (or more) entries (lines) in the .hidelist are concealed from the selection list.

To make the mechanism more expressive, backticks `...` within the entries can be used to moreover specify patterns: that is, as ECMAScript regular expressions.

  https://www.cplusplus.com/reference/regex/ECMAScript/

For instance, providing a .hidelist with the content

`.*\.mra`

or, alternatively,

`.*`.mra

inside the /media/fat/_Arcade root folder would hide all .mra files, so that the user can easily establish their own organisation of MRAs in sub-folders via symbolic links, while keeping /media/fat/_Arcade as the folder where MRAs are downloaded to, e.g. by the various updater scripts. This is just one example of the feature being useful for cosmetic tuning; I presume there are other use cases. Note that if using regular expressions via `...`, special characters such as ".", "*", etc. need to be suitably escaped (outside those quotes this is not necessary).

I have already implemented this making changes to file_io.cpp, it is not much effort and code. It may just require a bit more testing and perhaps a code review. Also, currently it does not work for zipped folders (@TODO). I am happy to commit it to a suitable branch of the Main_MiSTer repository.

Let me know what you think and if you believe this could be a desirable addition to the MiSTer.

Best wishes,
Frank

**Comment by Frank-Zeyda** [comment]
Please note: due to mark-down, the above did not show the back-tick symbols around the areas shaded in grey.

**Comment by sorgelig** [comment]
well, processing such files, especially with regexps will require additional CPU power, so it may interfere the core which may request some data, but HPS won't be able to reply timely. 
Anyway it can be considered. May be you can show your changes, so i can see how much changes it needs.

If your example of MRA hiding, it's not a big deal to have original MRAs while have subfolders with links. Subfolders are always shown on top.

**Comment by sorgelig** [comment]
after some thinking... I don't see much use of regexp.. Probably it's enough to list file extensions you want to hide. Or even empty file .nomedia (like on Android) would be enough.

**Comment by Frank-Zeyda** [comment]
Dear Alexey,

Thanks very much for the quick reply and consideration of the feature.

I presume that additional CPU power is only used when the directory is **once entered and read** by the OSD software - e.g. not on a periodic basis. I was not aware that the HPS is so critical to the behaviour of the core(s), but if that is the case I could offer to do some profiling, e.g. scanning a directory with 10,000 files and 1000 entries for patterns in the .hidelist and/or .showlist and see what difference that makes. My hunch is that regular expressions, when compiled, can be matched fairly efficiently (meaning in linear time via deterministic FSMs). But in my current, somewhat unoptimised, implementation, we would nonetheless require n*m matching operations if n is the number of files and m the number of patterns.  But there is scope for optimisation.

The change is not very invasive, we are looking at about 150 lines of code, profiting from the C++ std::regex class. How would I best share this with you? E.g. feel free to contact me via frank.zeyda@web.de if so, and I shall send a quick reply with the code. If you then think it is over the top, not a problem at all. Otherwise, I am happy to do some profiling to mitigate concerns that this could cause run-time issues. Also, when the .hidelist and .showlist files are not present, there is next to no penalty on run-time at all. I would only foresee a penalty when large lists are used, but am right now careful to venture such predictions without some degree of profiling.

> In your example of MRA hiding, it's not a big deal to have original MRAs
> while have subfolders with links. Subfolders are always shown on top.

Yes. We are talking about cosmetics here. It is perhaps not the best example of a use case... But there might be more interesting use cases... 

> Probably it's enough to list file extensions you want to hide. Or even empty file .nomedia (like on Android) would be enough.

Surely, that would likewise address the above use case. However, the feature as proposed, would of course give more flexibility to the users, providing there is *general interest* for it.

Best wishes,
Frank

**Comment by birdybro** [comment]
>I was not aware that the HPS is so critical to the behaviour of the core(s)

I could be wrong, but I think a good example of this that you can see for yourself, load any MegaCD game that uses saves, save the game, and open the OSD to do a save. The game lags for a bit until the save is updated. This likely demonstrates the OSD and HPS having some kind of relationship that is resource critical. You could probably use similar testing with more HPS-side intense cores like MegaCD to test your changes.

**Comment by sorgelig** [comment]
I don't see much use of regexp. May be you don't like MRA organization and it's got updated periodically, so i understand your motivation to hide a whole list and create subfolders with your organization.  So for this purpose simple .nomedia will be fine and it won't require any CPU power to process. Other folders with ROMs are filled by you, so it's up to you to sort and filter it once.
It's not like simple "add and forget". It will be maintained according to further changes, so i don't want to maintain features rarely used.

As for showing your changes - it's easy. You can fork this repo and apply your changes there.

**Comment by Frank-Zeyda** [comment]
Okay, in that case a fork may be the best solution to publish the change nonetheless for those that might find it useful, with a reference to the Main_MiSTer repository. I will post a link here once I manage to do so, and will endeavour to keep the fork in sync with developments on main master.

PS: I could independently look into implementing the **.nomedia** feature that you propose. That would, as you mentioned, address the motivating use case I outlined above and not take much work... Will post an update once this is done!

Best wishes,
Frank

**Comment by Frank-Zeyda** [comment]
Dear friends of MiSTer,

I made an implementation of the above feature available through:

[https://github.com/Frank-Zeyda/MiSTer_Custom_Zeyda](https://github.com/Frank-Zeyda/MiSTer_Custom_Zeyda)

This still requires (more) testing and, as said, does not work with zipped folders yet. But anyone who wants to try it, feel free - the fork is currently in sync with the forked upstream [Main_MiSTer](https://github.com/MiSTer-devel/Main_MiSTer), and I will endeavour to keep it like that.

@sorgelig: Sadly, I did not have time yet to test the **.nomedia** feature, though it is already implemented on my local machine... I will integrate the code and push it into the above [MiSTer_Custom_Zeyda](https://github.com/Frank-Zeyda/MiSTer_Custom_Zeyda) fork as soon as that's done, so that - if desired - you can quickly transfer the respective code. Will post a message here then.

Best wishes,
Frank

**Comment by Frank-Zeyda** [comment]
Dear Alexey and friends of MiSTer,

First of all, apologies for the very long radio silence — back in August 2021 I promised to post an update here "once this is done", and it then slipped under the radar for almost five years. Better late than never, I hope, as I have recently found time to complete and modernize the work.

**What's new**

- The fork has been rebased onto the current `master` and the repository has been renamed to better reflect its purpose: it now lives at https://github.com/Frank-Zeyda/Main_MiSTer-osd-filter (the old *MiSTer_Custom_Zeyda* link earlier in this thread still redirects there).
- The **`.nomedia`** marker file that you suggested, Alexey, is now implemented alongside `.showlist`/`.hidelist`: an (empty) `.nomedia` in a directory hides all regular files of that directory from the OSD browser, while subfolders remain visible — handy e.g. for `_Arcade`, where updater scripts keep re-populating the MRA list, so one can maintain their own organisation in subfolders. Only the file's existence matters, so this variant involves no parsing at all.
- The list files gained a few practical refinements: `#`/`;` comment lines, tolerance of Windows line endings and a UTF-8 BOM, the parent entry `..` is never filtered (so navigation cannot break), and invalid patterns are reported in the MiSTer log rather than silently ignored.
- The implementation was reworked to use the code-base's own facilities (`fileTextReader`/`FileReadLine`, `FileExists`) and follows the file's formatting conventions. The footprint is deliberately small: **155 added lines in `file_io.cpp`, no existing line modified**, hooked into `ScanDirectory()` at two short points. The fork's README documents the exact semantics.

**State of verification** *(updated after further hardening of the test suite)*

The parsing and filtering logic is covered by a documented host-side unit-test suite — 84 checks, including property-based tests — that compiles and exercises the very functions shipped in `file_io.cpp` (they are extracted verbatim from the source at build time, so the tests cannot drift from the shipped code). The suite holds 100% line and branch coverage of those helpers, runs clean under AddressSanitizer/UBSan, and continuous integration enforces all of this on every push, alongside a warning-free cross-build of the full binary with the official toolchain from the repository's devcontainer. What is still outstanding is testing on target hardware — rather than delay this update any further, I have deferred that step until it is clear whether there is interest in upstream integration (see below).

**Regarding the CPU-cost concern (HPS responsiveness)**

Your main reservation in 2021 was additional HPS load interfering with cores, so the implementation is shaped accordingly:

- Nothing runs periodically or in the background — the files are only consulted when the OSD scans a directory (opening the browser / entering a folder), on the same code path that already `stat()`s and sorts every entry.
- For everyone who does not opt in (no such files on the card), the entire overhead is a few cheap file-existence checks per directory scan — no parsing, no regex.
- `.nomedia` alone is exactly the zero-cost variant you proposed: a single existence check.
- Regex work only occurs for directories that actually contain a `.showlist`/`.hidelist`; patterns are compiled once per scan and matching is proportional to (entries × patterns), which for typical hand-written lists is negligible next to the per-entry `stat()` calls the scan already performs. I would be happy to profile a pathological case (e.g. a 10,000-file directory with a large list) on the DE10-Nano and post the numbers.

**How to proceed**

If you see a path towards integrating this into the main code-base — be it the full feature or only `.nomedia`, which you favoured back then — I will complete the testing on target, run the profiling above, and submit a PR against whichever branch you prefer. Otherwise, no hard feelings at all: the feature will simply continue to live in the fork, kept in sync with `master`, and I think this issue could then reasonably be closed.

Best wishes,
Frank
